src/Microsoft.AspNetCore.HostFiltering/HostFilteringMiddleware.cs

```diff
@@ -0,0 +1,182 @@
+                }
+                return Task.CompletedTask;
+            }
+
+            return _next(context);
+        }
+
+        private void EnsureConfigured()
+        {
+            if (_allowAnyNonEmptyHost == true || _allowedHosts?.Count > 0)
+            {
+                return;
+            }
+
+            var allowedHosts = new List<StringSegment>();
+            if (_options.AllowedHosts?.Count > 0)
+            {
+                if (!TryProcessHosts(_options.AllowedHosts, allowedHosts))
+                {
+                    _logger.LogDebug("Wildcard detected, all requests with hosts will be allowed.");
+                    _allowAnyNonEmptyHost = true;
+                    return;
+                }
+            }
+            else
+            {
+                if (!TryProcessHosts(_serverAddresses.Addresses.Select(address => BindingAddress.Parse(address).Host), allowedHosts))
+                {
+                    _logger.LogDebug("Wildcard detected, all requests with hosts will be allowed.");
+                    _allowAnyNonEmptyHost = true;
+                    return;
+                }
+            }
+
+            if (allowedHosts.Count == 0)
+            {
+                throw new InvalidOperationException("No allowed hosts were configured and none could be discovered from the server.");
+            }
+
+            _logger.LogDebug("Allowed hosts: " + string.Join("; ", allowedHosts));
+            _allowedHosts = allowedHosts;
+        }
+
+        // returns false if any wildcards were found
+        private bool TryProcessHosts(IEnumerable<string> incoming, IList<StringSegment> results)
+        {
+            foreach (var entry in incoming)
+            {
+                // Punycode. Http.Sys requires you to register Unicode hosts, but the headers contain punycode.
+                var host = new HostString(entry).ToUriComponent();
+
+                if (IsTopLevelWildcard(host))
+                {
+                    // Disable filtering
+                    return false;
+                }
+
+                if (!results.Contains(host, StringSegmentComparer.OrdinalIgnoreCase))
+                {
+                    results.Add(host);
+                }
+            }
+
+            return true;
+        }
+
+        private bool IsTopLevelWildcard(string host)
+        {
+            return (string.Equals("*", host, StringComparison.Ordinal) // HttpSys wildcard
+                           || string.Equals("[::]", host, StringComparison.Ordinal) // Kestrel wildcard, IPv6 Any
+                           || string.Equals("0.0.0.0", host, StringComparison.Ordinal)); // IPv4 Any
+        }
+
+        // This does not duplicate format validations that are expected to be performed by the host.
+        private bool CheckHost(HttpContext context)
+        {
+            var host = new StringSegment(context.Request.Headers[HeaderNames.Host].ToString()).Trim();
```

Comment:
As a general question for host validation, do we want to check the request line for the host?

Reply:
We don't have access to that, only the server sees it.